### PR TITLE
Extract the certificate validation from main and use a context

### PR DESCRIPTION
**Pull request by matrixise**

### CODE REVIEW

**Comment by JulienPalard** [comment]
Thanks for the PR @matrixise!

I'm not a huge fan of the `Context` object, I have a few ideas, feel free to tell me if you don't think it's better:

- Don't store `now()` in the context, just use `datetime.now()` every time, I know it's a context switch, but who cares?
- Don't store `verbose` in the context: use `logging.debug()` instead, and set loglevel to debug if `-v` is given.
- Raise in case of certificate error instead of returning a string.
- Give the limit as a `limit` parameter, as now the context only contains a limit.

I think it make it easier to use:

```python
print(validate_certificate(Context(now=datetime.utcnow(), limit=timedelta(days=14), verbose=args.verbose), Service("mdk.fr")))
```

to:

```python
try:
    validate_certificate(Service("mdk.fr"), limit=timedelta(days=14))
except CertificateValidationError as err:
    print(err)
```

what do you think?

**Comment by matrixise** [comment]
Question, when can we use the `args.verbose`, because I am not sure if this one is really useful.

**Comment by JulienPalard** [comment]
I mostly used it while developping, now while demoing the app lib: it's not really interesting to see nothing, you can't tell if it works or not, so running with `--version` gives you all expirations dates, prooving the thing does something.

File: certificate_watcher.py [hunk 1]
```diff
@@ -69,6 +70,28 @@ def __repr__(self):
...
 
+class CertificateValidationError(Exception):
+    pass
+
+
+def validate_certificate(service: Service, limit: timedelta):
+    try:
+        cert = get_server_certificate(service)
+    except socket.timeout:
+        raise CertificateValidationError(f"{service}: connect timeout")
```
Comment (JulienPalard) [inline certificate_watcher.py]:
```suggestion
    except socket.timeout as err:
        raise CertificateValidationError(f"{service}: connect timeout") from err
```

Else Python will display "During the handling of the error, another error occured", which is misleading. With the raise it'll display something like "The exception is the cause of the other one", which is right.

File: certificate_watcher.py [hunk 2]
```diff
@@ -69,6 +70,28 @@ def __repr__(self):
...
+    except ConnectionResetError:
+        raise CertificateValidationError(f"{service}: Connection reset")
```
Comment (JulienPalard) [inline certificate_watcher.py]:
```suggestion
    except ConnectionResetError as err:
        raise CertificateValidationError(f"{service}: Connection reset") from err
```

File: certificate_watcher.py [hunk 3]
```diff
@@ -69,6 +70,28 @@ def __repr__(self):
...
+    except Exception as err:
+        raise CertificateValidationError(f"{service}: {err!s}")
```
Comment (JulienPalard) [inline certificate_watcher.py]:
```suggestion
    except Exception as err:
        raise CertificateValidationError(f"{service}: {err!s}") from err
```

**Comment by JulienPalard** [comment]
I'm ~ok to drop --verbose, it's most very usefull, we can always add it back if needed, but if so please remove it from argparse.

**Review by JulienPalard — APPROVED** [review]

**Comment by JulienPalard** [comment]
Thanks for the PR @matrixise !

**Pull request merged**